TileStache/Goodies/VecTiles/transform.py

```diff
@@ -2067,47 +2137,51 @@ def remove_duplicate_features(
-        return distance * float(1 << (zoom + 8)) / 40075016.68
-
-    new_features = []
-    for feature in layer['features']:
-        shape, props, fid = feature
-
-        keep_feature = True
-        if shape.geom_type in geometry_types:
-            key = tuple([props.get(k) for k in property_keys])
-            seen_geoms = seen_items.get(key)
-
-            if seen_geoms is None:
-                # first time we've seen this item, so keep it in
-                # the output.
-                seen_items[key] = [shape]
+    deduplicator = _Deduplicator(property_keys, min_distance,
+                                 none_means_unique)
+
+    for source_layer in source_layers:
+        layer_index = -1
+        # because this post-processor can potentially modify
+        # multiple layers, and that wasn't how the return value
+        # system was designed, instead it modifies layers
+        # *in-place*. this is abnormal, and as such requires a
+        # nice big comment like this!
+        for index, feature_layer in enumerate(feature_layers):
+            layer_datum = feature_layer['layer_datum']
+            layer_name = layer_datum['name']
+            if layer_name == source_layer:
+                layer_index = index
+                break
+
+        if layer_index < 0:
+            # TODO: warn about missing layer when we get the
+            # ability to log.
+            continue
 
-            else:
-                # if the distance is greater than the minimum set
-                # for this zoom, then we also keep it.
-                distance = min([shape.distance(s) for s in seen_geoms])
+        layer = feature_layers[layer_index]
 
-                # correct for zoom - we want visual distance, which
-                # means (pseudo) pixels.
-                distance = meters_to_pixels(distance)
+        new_features = []
+        for feature in layer['features']:
+            shape, props, fid = feature
+            keep_feature = True
 
-                if distance > min_distance:
-                    # keep this geom to suppress any other labels
-                    # nearby.
-                    seen_geoms.append(shape)
+            if geometry_types is not None and \
+               shape.geom_type in geometry_types:
+                keep_feature = deduplicator.keep_feature(feature)
 
-                else:
-                    keep_feature = False
+            if keep_feature:
+                new_features.append(feature)
 
-        if keep_feature:
-            new_features.append(feature)
+        # NOTE! modifying the layer *in-place*.
+        layer['features'] = new_features
+        feature_layers[index] = layer
 
-    layer['features'] = new_features
-    return layer
+    # returning None here would normally indicate that the
+    # post-processor has done nothing. but because this
+    # modifies the layers *in-place* then all the return
+    # value is superfluous.
+    return None
```

Comment:
Just thinking out loud, I wonder if there would be any potential efficiencies gained by consolidating all the post processors that create new label features into a single one. If so, we might want to consider that in the future.